src/d3d9/d3d9_vertex_declaration.cpp

```diff
@@ -205,6 +206,135 @@ namespace dxvk {
+      return D3DFVF_XYZRHW;
+    else if (element.Usage == D3DDECLUSAGE_BLENDWEIGHT && element.UsageIndex == 0) {
+      DWORD fvfRet = MapD3DDeclTypeFloatToFvfXYZBn(element.Type);
+      if (likely(fvfRet != 0))
+        return fvfRet;
+      else {
+        Logger::warn("D3D9VertexDecl::MapD3DDeclToFvf: Unsupported set of D3DDECLUSAGE_BLENDWEIGHT / D3DDECLTYPE_*  / UsageIndex");
+        return 0;
+      }
+    }
+    else if (element.Usage == D3DDECLUSAGE_BLENDINDICES && element.Type == D3DDECLTYPE_UBYTE4 && element.UsageIndex == 0)
+      return D3DFVF_XYZB1;
+    else if (element.Usage == D3DDECLUSAGE_NORMAL && element.Type == D3DDECLTYPE_FLOAT3 && element.UsageIndex == 0)
+      return D3DFVF_NORMAL;
+    else if (element.Usage == D3DDECLUSAGE_PSIZE && element.Type == D3DDECLTYPE_FLOAT1 && element.UsageIndex == 0)
+      return D3DFVF_PSIZE;
+    else if (element.Usage == D3DDECLUSAGE_COLOR && element.Type == D3DDECLTYPE_D3DCOLOR) {
+      switch (element.UsageIndex)
+      {
+      case 0:
+        return D3DFVF_DIFFUSE;
+      case 1:
+        return D3DFVF_SPECULAR;
+      default:;
+        Logger::warn("D3D9VertexDecl::MapD3DDeclToFvf: Unsupported set of D3DDECLUSAGE_COLOR / D3DDECLTYPE_D3DCOLOR / UsageIndex");
+        return 0;
+      }
+    }
+    else if (element.Usage == D3DDECLUSAGE_TEXCOORD && element.UsageIndex < 8) {
+      DWORD fvfRet = MapD3DDeclUsageTexCoordToFvfTexCoordSize(element, fvf, texCountPostUpdate);
+      if (likely(fvfRet != 0))
+        return fvfRet;
+      else {
+        Logger::warn("D3D9VertexDecl::MapD3DDeclToFvf: Unsupported set of D3DDECLUSAGE_TEXCOORD / D3DDECLTYPE_* / UsageIndex");
+        return 0;
+      }
+    }
+
+    Logger::warn("D3D9VertexDecl::MapD3DDeclToFvf: Unsupported set of D3DDECLUSAGE_* / D3DDECLTYPE_* / UsageIndex");
+    return 0;
+  }
+
+
+  DWORD D3D9VertexDecl::MapD3DDeclTypeFloatToFvfXYZBn(BYTE type) {
+
+    switch (type)
+    {
+    default:;
+      return 0;
+    case D3DDECLTYPE_FLOAT1:
+      return D3DFVF_XYZB1;
+    case D3DDECLTYPE_FLOAT2:
+      return D3DFVF_XYZB2;
+    case D3DDECLTYPE_FLOAT3:
+      return D3DFVF_XYZB3;
+    case D3DDECLTYPE_FLOAT4:
+      return D3DFVF_XYZB4;
+    }
+  }
+
+
+  DWORD D3D9VertexDecl::MapD3DDeclUsageTexCoordToFvfTexCoordSize(
+      const D3DVERTEXELEMENT9& element,
+      DWORD fvf,
+      DWORD& texCountPostUpdate) {
+
+    // Check if bits of format for current UsageIndex are free in the fvf
+    // It is necessary to skip multiple initializations of the bitfield because
+    // returned value is bitwise or-ed to final fvf DWORD.
+    if ((D3DFVF_TEXCOORDSIZE1(element.UsageIndex) & fvf) != 0)
+      return 0;
+
+    // Update max texture's index in fvf
+    DWORD lastMaxTexCount = (fvf >> 8) & 0x7;
+    DWORD currentTexCount = element.UsageIndex + 1;
+
+    if (lastMaxTexCount < currentTexCount)
+      texCountPostUpdate = currentTexCount << 8;
```

Comment:
This seems wrong because we don't OR in the `texCountPostUpdate` until after all this is done, yet we seem to try to pull it out here and check it?

Reply:
You mean that lastMaxTexCount will always be 0, right? 
Then we will always get the last counter instead of largest. 
Good point.
Thanks.

Reply:
New commit [e05092bc921937e4026d5196f7fd1bf9f23f9ee4](https://github.com/doitsujin/dxvk/pull/2924/commits/e05092bc921937e4026d5196f7fd1bf9f23f9ee4) fixes problem of texCountPostUpdate.
Additionally, function MapD3DDeclUsageTexCoordToFvfTexCoordSize may return 0 as valid value. 
Functionally it worked, but when 0 was treated as incorrect value then unwanted warning was generated.